internal/core/block/block.go

```diff
@@ -0,0 +1,361 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package coreblock
+
+import (
+	"bytes"
+	"sort"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/codec/dagcbor"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/multiformats/go-multicodec"
+
+	"github.com/sourcenetwork/defradb/internal/core"
+	"github.com/sourcenetwork/defradb/internal/core/crdt"
+)
+
+// Schema is the IPLD schema type that represents a `Block`.
+var (
+	Schema          schema.Type
+	SchemaPrototype schema.TypedPrototype
+)
+
+func init() {
+	Schema, SchemaPrototype = mustSetSchema(
+		&Block{},
+		&DAGLink{},
+		&CRDT{},
+		&crdt.LWWRegDelta{},
+		&crdt.CompositeDAGDelta{},
+		&crdt.CounterDelta[int64]{}, // Only need to call one of the CounterDelta types.
+	)
+}
+
+type schemaDefinition interface {
+	// IPLDSchemaBytes returns the IPLD schema representation for the type.
+	IPLDSchemaBytes() []byte
+}
+
+func mustSetSchema(schemas ...schemaDefinition) (schema.Type, schema.TypedPrototype) {
+	schemaBytes := make([][]byte, 0, len(schemas))
+	for _, s := range schemas {
+		schemaBytes = append(schemaBytes, s.IPLDSchemaBytes())
+	}
+
+	ts, err := ipld.LoadSchemaBytes(bytes.Join(schemaBytes, nil))
+	if err != nil {
+		panic(err)
+	}
+	blockSchemaType := ts.TypeByName("Block")
+
+	// Calling bindnode.Prototype here ensure that [Block] and all the types it contains
+	// are compatible with the IPLD schema defined by blockSchemaType.
+	// If [Block] and `blockSchematype` do not match, this will panic.
+	proto := bindnode.Prototype(&Block{}, blockSchemaType)
+
+	return blockSchemaType, proto
+}
+
+// DAGLink represents a link to another object in a DAG.
+type DAGLink struct {
+	// Name is the name of the link.
+	//
+	// This will be either the field name of the CRDT delta or "_head" for the head link.
+	Name string
+	// Link is the CID link to the object.
+	cidlink.Link
+}
+
+// IPLDSchemaBytes returns the IPLD schema representation for the DAGLink.
+//
+// This needs to match the [DAGLink] struct or [mustSetSchema] will panic on init.
+func (l DAGLink) IPLDSchemaBytes() []byte {
```

Comment:
This is already taken care of by a compile time check on line 37 of `block.go`. No need to worry about doing that for every type individually. It will automatically be taken care of for every new type that we add to `mustSetSchema`.